File: Problem_1 Trapping Rain_Water.py
# // Time Complexity : O(n)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode :
# // Any problem you faced while coding this :
#
#
# // Your code here along with comments explaining your approach

class Solution:
    def trap(self, height: List[int]) -> int:
        if height == None or len(height) == 0:
            return 0
        n = len(height)
        lmax = 0
        rmax = 0
        left = 0
        right = n - 1
        result = 0
        while left < right:
            if height[left] <= height[right]:
                if lmax <= height[left]:
                    lmax = height[left]
                else:
                    result = result + lmax - height[left]
                left = left + 1
            else:
                if rmax <= height[right]:
                    rmax = height[right]
                else:
                    result = result + rmax - height[right]
                right = right - 1
        return result

File: Problem1_TrappingRainWater.java
// Time Complexity :o(N)
// Space Complexity : o(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution {
    public int trap(int[] height) {
        if(height==null||height.length<=2)
            return 0;
        
        int lw=0,rw=0,l=0,r=height.length-1;
        int result=0;
        
        while(l<=r)
        {
            if(lw<=rw)
            {
                if(height[l]<lw)
                {
                    result+=(lw-height[l]);
                }
                else
                {
                    lw=height[l];
                }
                l++;
            }
            else
            {
                if(height[r]<rw)
                {
                    result+=(rw-height[r]);
                }
                else
                {
                    rw=height[r];
                }
                r--;
            }
        }
        return result;
    }
}